src/Microsoft.ML.TimeSeries/SrCnnEntireAnomalyDetector.cs

```diff
@@ -0,0 +1,858 @@
+            private double[] _magLogList;
+            private double[] _spectralList;
+            private double[] _transRe;
+            private double[] _transIm;
+            private double[] _ifftRe;
+            private double[] _ifftIm;
+            private double[] _ifftMagList;
+            private double[] _cumSumList;
+            private double[] _cumSumShift;
+            private double[] _zeroArray;
+            //used in AnomalyAndExpectedValue and AnomalyAndMargin
+            private double[] _deAnomalyData;
+            //used in AnomalyAndMargin mode
+            private double[] _units;
+            private double[] _val;
+            private double[] _trends;
+            private double[] _curWindow;
+
+            public SrCnnEntireModeler(double threshold, double sensitivity, SrCnnDetectMode detectMode)
+            {
+                _threshold = threshold;
+                _sensitivity = sensitivity;
+                _detectMode = detectMode;
+                _predictArray = new double[_lookaheadWindowSize + 1];
+            }
+
+            public void Train(double[] values, ref double[][] results)
+            {
+                if (results == null)
+                {
+                    results = new double[values.Length][];
+                    for (int i = 0; i < results.Length; ++i)
+                    {
+                        results[i] = new double[_outputLengthArray[(int)_detectMode]];
+                    }
+                }
+                else if (results.Length > values.Length)
+                {
+                    Array.Resize<double[]>(ref results, values.Length);
+                }
+                SpectralResidual(values, results, _threshold);
+                //Optional Steps
+                if (_detectMode == SrCnnDetectMode.AnomalyAndMargin)
+                {
+                    GetMargin(values, results, _sensitivity);
+                }
+                else if (_detectMode == SrCnnDetectMode.AnomalyAndExpectedValue)
+                {
+                    GetExpectedValue(values, results);
+                }
+            }
+
+            private void AllocateDoubleArray(ref double[] arr, int length)
+            {
+                if (arr == null)
+                {
+                    arr = new double[length];
+                }
+                else if (arr.Length != length)
+                {
+                    Array.Resize<double>(ref arr, length);
+                }
+            }
+
+            private void AllocateZeroArray(ref double[] arr, int length)
+            {
+                if (arr == null)
+                {
+                    arr = new double[length];
+                }
+                else if (arr.Length != length)
+                {
+                    Array.Resize<double>(ref arr, length);
+                }
+                for (int i = 0; i < arr.Length; ++i)
+                {
+                    arr[i] = 0.0;
+                }
```

Comment:
Values will be 0 by default.

Reply:
Since values are 0 by default, removed this fuction, use AllocateDoubleArray instead, they are equally.

---
In reply to: [428513049](https://github.com/dotnet/machinelearning/pull/5135#discussion_r428513049) [](ancestors = 428513049)